azure-client-runtime/src/main/java/com/microsoft/azure/PagedList.java

```diff
@@ -139,70 +140,122 @@ protected void setCurrentPage(Page<E> currentPage) {
      * The implementation of {@link ListIterator} for PagedList.
      */
     private class ListItr implements ListIterator<E> {
-        /** The list iterator for the actual list of items. */
-        private ListIterator<E> itemsListItr;
+        /**
+         * index of next element to return.
+         */
+        private int nextIndex;
+        /**
+         * index of last element returned; -1 if no such action happened.
+         */
+        private int lastRetIndex = -1;
 
         /**
          * Creates an instance of the ListIterator.
          *
          * @param index the position in the list to start.
          */
         ListItr(int index) {
-            itemsListItr = items.listIterator(index);
+            this.nextIndex = index;
         }
 
         @Override
         public boolean hasNext() {
-            return itemsListItr.hasNext() || hasNextPage();
+            return this.nextIndex != items.size() || hasNextPage();
         }
 
         @Override
         public E next() {
-            if (!itemsListItr.hasNext()) {
+            if (this.nextIndex >= items.size()) {
                 if (!hasNextPage()) {
                     throw new NoSuchElementException();
                 } else {
-                    int size = items.size();
                     loadNextPage();
-                    itemsListItr = items.listIterator(size);
+                }
+                return next();
+            } else {
+                try {
+                    E nextItem = items.get(this.nextIndex);
+                    this.lastRetIndex = this.nextIndex;
+                    this.nextIndex = this.nextIndex + 1;
+                    return nextItem;
+                } catch (IndexOutOfBoundsException ex) {
+                    throw new ConcurrentModificationException();
                 }
             }
-            return itemsListItr.next();
         }
 
         @Override
         public void remove() {
-            itemsListItr.remove();
+            if (this.lastRetIndex < 0) {
+                throw new IllegalStateException();
+            } else {
+                try {
+                    items.remove(this.lastRetIndex);
+                    this.nextIndex = this.lastRetIndex;
+                    this.lastRetIndex = -1;
+                } catch (IndexOutOfBoundsException ex) {
+                    throw new ConcurrentModificationException();
+                }
+            }
         }
 
         @Override
         public boolean hasPrevious() {
-            return itemsListItr.hasPrevious();
+            return this.nextIndex != 0;
         }
 
         @Override
         public E previous() {
-            return itemsListItr.previous();
+            int i = this.nextIndex - 1;
+            if (i < 0) {
+                throw new NoSuchElementException();
+            } else {
+                if (i >= items.size()) {
```

Comment:
Simplify this line by merging it with the previous line (`else` becomes an `else if`).

Reply:
thanks @daschult , fixed.